src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
+  // i.e., the loop guard was satisfied.
+  const auto entered_loop =
+    new_tmp_symbol(
+      bool_typet(),
+      loop_head_location,
+      mode,
+      symbol_table,
+      std::string(ENTERED_LOOP) + "__" + std::to_string(loop_number))
+      .symbol_expr();
+  pre_loop_head_instrs.add(
+    goto_programt::make_decl(entered_loop, loop_head_location));
+  pre_loop_head_instrs.add(
+    goto_programt::make_assignment(entered_loop, false_exprt{}));
+
+  // Create a snapshot of the invariant so that we can check the base case,
+  // if the loop is not vacuous and must be abstracted with contracts.
+  const auto initial_invariant_val =
+    new_tmp_symbol(
+      bool_typet(), loop_head_location, mode, symbol_table, INIT_INVARIANT)
+      .symbol_expr();
+  pre_loop_head_instrs.add(
+    goto_programt::make_decl(initial_invariant_val, loop_head_location));
+  {
+    // Although the invariant at this point will not have side effects,
+    // it is still a C expression, and needs to be "goto_convert"ed.
+    // Note that this conversion may emit many GOTO instructions.
+    code_assignt initial_invariant_value_assignment{
+      initial_invariant_val, invariant};
+    initial_invariant_value_assignment.add_source_location() =
+      loop_head_location;
+    converter.goto_convert(
+      initial_invariant_value_assignment, pre_loop_head_instrs, mode);
+  }
+
+  // Create a temporary variable to track base case vs inductive case
+  // instrumentation of the loop.
+  const auto in_base_case =
+    new_tmp_symbol(
+      bool_typet(), loop_head_location, mode, symbol_table, "__in_base_case")
+      .symbol_expr();
+  pre_loop_head_instrs.add(
+    goto_programt::make_decl(in_base_case, loop_head_location));
+  pre_loop_head_instrs.add(
+    goto_programt::make_assignment(in_base_case, true_exprt{}));
+
+  // TODO: __populate(w_loop_i, <loop_assigns>);
+
+  goto_function.body.destructive_insert(
+    loop_head, add_pragma_disable_assigns_check(pre_loop_head_instrs));
+
+  //                        |    STEP:
+  //                  --.   |      assert (initial_invariant_val);
+  //                    |   |      check_assigns_inclusion(w_loop, w_parent);
+  // loop assigns check |   |      in_base_case = false;
+  //  - not applicable   >=======  in_loop_havoc_block = true;
+  // func assigns check |   |      havoc (assigns_set);
+  //  + deferred        |   |      in_loop_havoc_block = false;
+  //                  --'   |      assume (invariant_expr);
+  //                        `-     old_variant_val = decreases_clause_expr;
+
+  goto_programt step_instrs;
+
+  // set of targets to havoc
+  assignst to_havoc;
+
+  if(assigns_clause.is_nil())
+  {
+    // No assigns clause was specified for this loop.
+    // Infer memory locations assigned by the loop from the loop instructions
+    // and the inferred aliasing relation.
+    try
+    {
+      get_assigns_dfcc(local_may_alias, goto_function, idx, to_havoc);
+
```

Comment:
@qinheping now that we are deprecating `__CPROVER_POINTER_OBJECT` in assigns clauses this function needs to widen targets using `__CPROVER_object_whole`